libcaf_core/caf/flow/op/retry.hpp

```diff
@@ -0,0 +1,172 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/config.hpp"
+#include "caf/detail/assert.hpp"
+#include "caf/flow/backpressure_overflow_strategy.hpp"
+#include "caf/flow/observer.hpp"
+#include "caf/flow/op/hot.hpp"
+#include "caf/flow/subscription.hpp"
+
+#include <deque>
+#include <utility>
+
+namespace caf::flow::op {
+
+template <class T, class Predicate>
+class retry_sub : public subscription::impl_base, public observer_impl<T> {
+public:
+  // -- constructors, destructors, and assignment operators --------------------
+
+  retry_sub(coordinator* parent, observer<T> out, Predicate predicate,
+            observable<T> in)
+    : parent_(parent),
+      out_(std::move(out)),
+      in_(std::move(in)),
+      predicate(std::move(predicate)) {
+    // nop
+  }
+
+  // -- implementation of subscription -----------------------------------------
+
+  coordinator* parent() const noexcept override {
+    return parent_;
+  }
+
+  bool disposed() const noexcept override {
+    return !out_;
+  }
+
+  void request(size_t new_demand) override {
+    if (new_demand == 0)
+      return;
+    demand_ += new_demand;
+  }
+
+  // -- implementation of observer_impl ----------------------------------------
+
+  void ref_coordinated() const noexcept override {
+    ref();
+  }
+
+  void deref_coordinated() const noexcept override {
+    deref();
+  }
+
+  void on_subscribe(subscription sub) override {
+    if (sub_) {
+      sub_.cancel();
+      sub_.release_later();
+    }
+    sub_ = std::move(sub);
+    sub_.request(256);
```

Comment:
The bookkeeping of the demand looks broken to me. Our subscriber communicates demand to use, that we should forward to `sub_`. If `sub_` goes down and re-subscribe, the existing demand should be passed over. There's also a `sub_.request(1);` further down below that I think should not exist.

The subscriber should drive demand. The `retry` operator should not hand out demand on its own, since it's not buffering items.